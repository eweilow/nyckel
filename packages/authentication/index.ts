export { decodeJWT } from "./jwt/decode";

export { verifyAndDecodeJWT, DecodedJWT } from "./jwt/verify";

export { createRateLimiter } from "./utils/rateLimiter";

export { concatUrl } from "./utils/concatUrl";
export { getSafeHostname } from "./utils/safeHostname";

export { verifyCSRFPair } from "./state/verify";
export { CSRFPair } from "./state/pair";

export {
  createGlobalAuthenticationConfig,
  GlobalAuthenticationConfig
} from "./utils/globalConfig";

export { logoutUser } from "./methods/logout";
export { authorizeUser } from "./methods/authorize";
export { requestToken } from "./methods/token";
export { getUserInfo, UserInfo } from "./methods/profile";
export { attemptToRefreshToken } from "./methods/refresh";
