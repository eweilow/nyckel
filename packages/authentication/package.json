{
  "name": "@nyckel/authentication",
  "version": "0.0.1-alpha.11",
  "description": "A framework agnostic library for Auth0 authentication in hybrid web apps",
  "main": "./dist/cjs",
  "module": "./dist/esm",
  "publishConfig": {
    "access": "public"
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/eweilow/nyckel.git"
  },
  "author": "eweilow",
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/eweilow/nyckel/issues"
  },
  "homepage": "https://github.com/eweilow/nyckel",
  "dependencies": {
    "csrf": "^3.1.0",
    "jsonwebtoken": "^8.5.1",
    "jwks-rsa": "^1.6.0",
    "lru-cache": "^5.1.1",
    "node-fetch": "^2.6.0"
  },
  "devDependencies": {
    "@types/jsonwebtoken": "8.3.3",
    "@types/lru-cache": "5.1.0",
    "@types/node": "12.7.1",
    "@types/node-fetch": "2.5.0"
  },
  "keywords": [
    "nyckel"
  ],
  "types": "./",
  "gitHead": "48ceeacb01d90f875aa79ad1c6cfc593184f579b"
}
