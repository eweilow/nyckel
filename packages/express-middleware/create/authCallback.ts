import { RequestHandler } from "express";
import { NyckelExpressMiddlewareOptions } from "../middleware";
import { asyncHandler } from "..";
import {
  concatUrl,
  requestToken,
  verifyCSRFPair
} from "@nyckel/authentication";

export function createCallbackHandler(
  opts: NyckelExpressMiddlewareOptions
): RequestHandler {
  const { authConfig } = opts;

  return asyncHandler(async (req, res) => {
    if (!req.query.state) {
      throw new Error("Expected query to contain state!");
    }
    if (!req.query.code) {
      throw new Error("Expected query to contain code!");
    }

    const session = await req.session.get();

    try {
      const { csrfSecret } = session;
      await verifyCSRFPair({
        secret: csrfSecret,
        token: req.query.state
      });
    } catch (err) {
      await req.session.delete();
      throw err;
    }

    const data = await requestToken(
      req.query.code,
      concatUrl(req.realHost, "/auth/callback"),
      authConfig
    );
    await req.session.set(data);
    res.redirect("/");
  });
}
