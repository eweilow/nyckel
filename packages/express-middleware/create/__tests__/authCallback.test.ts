import { createCallbackHandler } from "../authCallback";
import { verifyCSRFPair, requestToken } from "@nyckel/authentication";

const requestTokenData = { requestTokenData: "yes" };
jest.mock("@nyckel/authentication", () => {
  const orig = jest.requireActual("@nyckel/authentication");
  return {
    ...orig,
    requestToken: jest.fn(async () => requestTokenData),
    verifyCSRFPair: jest.fn(async ({ secret, token }) => {
      if (secret === "secret" && token === "state") {
        return;
      }
      throw new Error("Unable to verify CSRF pair (test message)");
    })
  };
});

describe("createCallbackHandler", () => {
  it("returns a function", () => {
    expect(typeof createCallbackHandler({} as any)).toBe("function");
  });

  it("throws if state is not in query", async () => {
    const handler = createCallbackHandler({} as any);

    const next = jest.fn();
    await handler(
      {
        query: {}
      } as any,
      {} as any,
      next
    );

    expect(next).toBeCalledTimes(1);
    const error = next.mock.calls[0][0];
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toMatchSnapshot();
  });

  it("throws if code is not in query", async () => {
    const handler = createCallbackHandler({} as any);

    const next = jest.fn();
    await handler(
      {
        query: {
          state: "yes"
        }
      } as any,
      {} as any,
      next
    );

    expect(next).toBeCalledTimes(1);
    const error = next.mock.calls[0][0];
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toMatchSnapshot();
  });

  it("throws if csrfSecret is not in session", async () => {
    const handler = createCallbackHandler({} as any);

    const session = {
      get: jest.fn(async () => ({})),
      delete: jest.fn(async () => {})
    };

    const next = jest.fn();
    await handler(
      {
        query,
        session
      } as any,
      {} as any,
      next
    );

    expect(next).toBeCalledTimes(1);
    const error = next.mock.calls[0][0];
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toMatchSnapshot();
  });

  const query = {
    state: "state",
    code: "code"
  };

  it("attempts to verify the CSRF pair and deletes session if it cannot verify the pair", async () => {
    const handler = createCallbackHandler({} as any);

    const session = {
      get: jest.fn(async () => ({ csrfSecret: "invalidSecret" })),
      delete: jest.fn(async () => {})
    };

    const next = jest.fn();
    await handler(
      {
        query,
        session
      } as any,
      {} as any,
      next
    );

    expect(session.get).toBeCalledTimes(1);
    expect(verifyCSRFPair).toBeCalled();
    expect(session.delete).toBeCalledTimes(1);

    expect(next).toBeCalledTimes(1);
    const error = next.mock.calls[0][0];
    expect(error).toBeInstanceOf(Error);
    expect(error.message).toMatchSnapshot();
  });

  it("attempts to verify the CSRF pair and handles correctly if it can verify the pair", async () => {
    const handler = createCallbackHandler({ authConfig: "authConfig" } as any);

    const session = {
      get: jest.fn(async () => ({ csrfSecret: "secret" })),
      delete: jest.fn(async () => {}),
      set: jest.fn(async () => {})
    };

    const redirect = jest.fn();

    const next = jest.fn();
    await handler(
      {
        realHost: "http://example.com",
        query,
        session
      } as any,
      {
        redirect
      } as any,
      next
    );

    expect(session.get).toBeCalledTimes(1);
    expect(verifyCSRFPair).toBeCalled();
    expect(session.delete).not.toBeCalled();
    expect(next).not.toBeCalled();

    expect(requestToken).toBeCalledWith(
      "code",
      "http://example.com/auth/callback",
      "authConfig"
    );
    expect(session.set).toBeCalledWith(requestTokenData);

    expect(redirect).toBeCalledWith("http://example.com/");
  });
});
