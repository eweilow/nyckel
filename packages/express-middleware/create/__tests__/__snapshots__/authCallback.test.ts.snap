// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCallbackHandler attempts to verify the CSRF pair and deletes session if it cannot verify the pair 1`] = `"Unable to verify CSRF pair (test message)"`;

exports[`createCallbackHandler throws if code is not in query 1`] = `"Expected query to contain code!"`;

exports[`createCallbackHandler throws if csrfSecret is not in session 1`] = `"Session must contain a csrfSecret!"`;

exports[`createCallbackHandler throws if state is not in query 1`] = `"Expected query to contain state!"`;
